README.md
# SFTP file output plugin for Embulk
[![Build Status](https://travis-ci.org/embulk/embulk-output-sftp.svg)](https://travis-ci.org/embulk/embulk-output-sftp)

Stores files on a SFTP Server

## Overview

* **Plugin type**: file output
* **Load all or nothing**: no
* **Resume supported**: no
* **Cleanup supported**: no

## Configuration

- **host**: (string, required)
- **port**: (int, default: `22`)
- **user**: (string, required)
- **password**: (string, default: `null`)
- **secret_key_file**: (string, default: `null`) [see below](#secret-keyfile-configuration)
- **secret_key_passphrase**: (string, default: `""`)
- **user_directory_is_root**: (boolean, default: `true`)
- **timeout**: sftp connection timeout seconds (integer, default: `600`)
- **path_prefix**: Prefix of output paths (string, required)
- **file_ext**: Extension of output files (string, required)
- **sequence_format**: Format for sequence part of output files (string, default: `".%03d.%02d"`)
- **rename_file_after_upload**: Upload `file_ext` + ".tmp" first, then rename it after upload finish (boolean, default: `false`)
- **local_buffering**: Use local temp file to buffer records. If `false`, plugin will buffer records to remote file directly, with ".tmp" as filename suffix (boolean, default: `true`)
- **temp_file_threshold**: Maximum file size of local temp file, plugin will flush (append) to remote file when local temp file reaches threshold (long, default: `5368709120`, ie. 5GiB, min: 50MiB, max: 10GiB)

### Proxy configuration

- **proxy**:
    - **type**: (string(http | socks | stream), required, default: `null`)
        - **http**: use HTTP Proxy
        - **socks**: use SOCKS Proxy
        - **stream**: Connects to the SFTP server through a remote host reached by SSH
    - **host**: (string, required)
    - **port**: (int, default: `22`)
    - **user**: (string, optional)
    - **password**: (string, optional, default: `null`)
    - **command**: (string, optional)
    
## Example

```yaml
out:
  type: sftp
  host: 127.0.0.1
  port: 22
  user: civitaspo
  secret_key_file: /Users/civitaspo/.ssh/id_rsa
  secret_key_passphrase: secret_pass
  user_directory_is_root: false
  timeout: 600
  path_prefix: /data/sftp
  file_ext: _20151020.tsv
  sequence_format: ".%01d%01d"
  temp_file_threshold: 10737418240 # 10GiB
```

With proxy
```yaml
out:
  type: sftp
  host: 127.0.0.1
  port: 22
  user: embulk
  secret_key_file: /Users/embulk/.ssh/id_rsa
  secret_key_passphrase: secret_pass
  user_directory_is_root: false
  timeout: 600
  path_prefix: /data/sftp
  proxy:
    type: http
    host: proxy_host
    port: 8080
    user: proxy_user
    password: proxy_secret_pass
    command:
```

### Secret Keyfile configuration

Please set path of secret_key_file as follows.
```yaml
out:
  type: sftp
  ...
  secret_key_file: /path/to/id_rsa
  ...
```

You can also embed contents of secret_key_file at config.yml.
```yaml
out:
  type: sftp
  ...
  secret_key_file:
    content |
      -----BEGIN RSA PRIVATE KEY-----
      ABCDEFG...
      HIJKLMN...
      OPQRSTU...
      -----END RSA PRIVATE KEY-----
  ...
```

## Run Example

```
$ ./gradlew classpath
```

Use `vagrant` to start a remote sshd server:

```
$ vagrant up
```

Run:


```
$ embulk run -Ilib example/sample.yml
```

## Build

```
$ ./gradlew gem  # -t to watch change of files and rebuild continuously
$ ./gradlew bintrayUpload # release embulk-output-sftp to Bintray maven repo
```

## Note

This plugin uses "org.apache.commons:commons-vfs" and the library uses the logger "org.apache.commons.logging.Log". So, this plugin suppress the logger's message except when embulk log level is debug.

## Contributors
- Satoshi Akama (@sakama)
- Rudolph Miller (@Rudolph-Miller)
- Naotoshi Seo (@sonots)


CHANGELOG.md
0.2.5 (2020-08-10)
- Close unused Session.
  - https://github.com/embulk/embulk-output-sftp/pull/60
0.2.4 (2020-05-20)
- Add more log to monitor.
  - https://github.com/embulk/embulk-output-sftp/pull/58
0.2.2 (2019-07-26)
- Do not retry when "Connection refused" is returned and throw ConfigException
  - https://github.com/embulk/embulk-output-sftp/pull/56
0.2.1 (2018-10-23)
- Improved logic that detects exception is retryable or not
  - https://github.com/embulk/embulk-output-sftp/pull/52
0.2.0 (2018-08-29)
- Fix bug of stalling when closing `remoteFile`, this version is recommended over 0.1.11
  - https://github.com/embulk/embulk-output-sftp/pull/51
0.1.11 (2018-08-27)
- Enhance: Add 2 new configs `local_temp_file` (boolean) and `temp_file_threshold` (long)
  - https://github.com/embulk/embulk-output-sftp/pull/50
0.1.10 (2018-05-07)
- Fix: Use java.util.regex.Pattern for host name validation
  - https://github.com/embulk/embulk-output-sftp/pull/49
0.1.9 (2018-04-26)
- Enhance: Add validation for "host" and "proxy.host"
  - https://github.com/embulk/embulk-output-sftp/pull/48
0.1.8 (2018-03-20)
- Change input stream buffer size to 32MB to reduce number of storage reads
  - https://github.com/embulk/embulk-output-sftp/pull/47
0.1.7 (2018-03-14)
- Enhance: Upload with ".tmp" suffix and rename file name after upload 
  - https://github.com/embulk/embulk-output-sftp/pull/43
  - https://github.com/embulk/embulk-output-sftp/pull/44

0.1.6 (2018-03-13)
- Fix: Change input stream buffer size to 4MB to fix hanging issue
  - https://github.com/embulk/embulk-output-sftp/pull/46
0.1.5 (2018-03-13)
- Fix: Fix random hanging and log transfer progress
  - https://github.com/embulk/embulk-output-sftp/pull/45

0.1.4 (2017-12-21)
- Fix: Disable remote temporary file rename logic
  - https://github.com/embulk/embulk-output-sftp/pull/41
- Enhance: Upgrade "commons-vfs2", "com.jcraft:jsch" and "commons-io:commons-io"
  - https://github.com/embulk/embulk-output-sftp/pull/42

0.1.3 (2017-11-07)
- Enhance: Create temporary file and rename it after upload completed
  - https://github.com/embulk/embulk-output-sftp/pull/37

0.1.2 (2017-07-11)
- Fix: Increment fileIndex after complete uploadFile
  - https://github.com/embulk/embulk-output-sftp/pull/35

0.1.1 (2017-05-29)
- Fix: Improve retry logic
  - https://github.com/embulk/embulk-output-sftp/pull/34

0.0.9 (2017-03-09)
==================
- Fix: Hide password in the log
  - https://github.com/civitaspo/embulk-output-sftp/pull/25

0.0.8 (2016-09-26)
==================
- Fix: Use second as timetout setting instead of milli second
  - https://github.com/civitaspo/embulk-output-sftp/pull/22
- Fix: Fix CI failure only with Java7
  - https://github.com/civitaspo/embulk-output-sftp/pull/21
- Fix: Format code that were warned by `./gradlew checkstyle` command
  - https://github.com/civitaspo/embulk-output-sftp/pull/23

0.0.7 (2016-03-22)
==================
- Fix: Plugin throws ClassNotFoundException with EmbulkEmbed
  - https://github.com/civitaspo/embulk-output-sftp/pull/19

0.0.6 (2016-03-16)
==================
- Fix: Avoid some connection errors
  - https://github.com/civitaspo/embulk-output-sftp/pull/17

0.0.5 (2016-03-09)
==================
- Add: Support MapReduce executor
  - https://github.com/civitaspo/embulk-output-sftp/pull/8
- Fix: Use ConfigException instead of RuntimeException
  - https://github.com/civitaspo/embulk-output-sftp/pull/9/files
- Fix: Check to exist parent directory before uploading files
  - https://github.com/civitaspo/embulk-output-sftp/pull/13
- Add: Support proxy settings
  - https://github.com/civitaspo/embulk-output-sftp/pull/11


src/main/java/org/embulk/output/sftp/SftpFileOutputPlugin.java
package org.embulk.output.sftp;

import org.embulk.config.Config;
import org.embulk.config.ConfigDefault;
import org.embulk.config.ConfigDiff;
import org.embulk.config.ConfigSource;
import org.embulk.config.Task;
import org.embulk.config.TaskReport;
import org.embulk.config.TaskSource;
import org.embulk.spi.Exec;
import org.embulk.spi.FileOutputPlugin;
import org.embulk.spi.TransactionalFileOutput;
import org.embulk.spi.unit.LocalFile;
import org.slf4j.Logger;

import javax.validation.constraints.Max;
import javax.validation.constraints.Min;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public class SftpFileOutputPlugin
        implements FileOutputPlugin
{
    private final Logger logger = Exec.getLogger(SftpFileOutputPlugin.class);

    public interface PluginTask
            extends Task
    {
        @Config("host")
        String getHost();

        @Config("port")
        @ConfigDefault("22")
        int getPort();

        @Config("user")
        String getUser();

        @Config("password")
        @ConfigDefault("null")
        Optional<String> getPassword();

        @Config("secret_key_file")
        @ConfigDefault("null")
        Optional<LocalFile> getSecretKeyFilePath();
        void setSecretKeyFilePath(Optional<LocalFile> secretKeyFilePath);

        @Config("secret_key_passphrase")
        @ConfigDefault("\"\"")
        String getSecretKeyPassphrase();

        @Config("user_directory_is_root")
        @ConfigDefault("true")
        boolean getUserDirIsRoot();

        @Config("timeout")
        @ConfigDefault("600") // 10 minutes
        int getSftpConnectionTimeout();

        @Config("max_connection_retry")
        @ConfigDefault("5") // 5 times retry to connect sftp server if failed.
        int getMaxConnectionRetry();

        @Config("path_prefix")
        String getPathPrefix();

        @Config("file_ext")
        String getFileNameExtension();

        @Config("sequence_format")
        @ConfigDefault("\"%03d.%02d.\"")
        String getSequenceFormat();

        @Config("proxy")
        @ConfigDefault("null")
        Optional<ProxyTask> getProxy();

        @Config("rename_file_after_upload")
        @ConfigDefault("false")
        boolean getRenameFileAfterUpload();

        // if `false`, plugin will use remote file as buffer
        @Config("local_buffering")
        @ConfigDefault("true")
        boolean getLocalBuffering();

        @Min(50L * 1024 * 1024) // 50MiB
        @Max(10L * 1024 * 1024 * 1024) // 10GiB
        @Config("temp_file_threshold")
        @ConfigDefault("5368709120") // 5GiB
        long getTempFileThreshold();
    }

    @Override
    public ConfigDiff transaction(ConfigSource config, int taskCount,
            FileOutputPlugin.Control control)
    {
        PluginTask task = config.loadConfig(PluginTask.class);
        SftpUtils sftpUtils = null;
        try {
            sftpUtils = new SftpUtils(task);
            sftpUtils.validateHost(task);
            if (task.getSecretKeyFilePath().isPresent()) {
                logger.info("checksum of key file: {}", sftpUtils.getChecksum(task.getSecretKeyFilePath().get().getPath()));
            }
        }
        finally {
            if (sftpUtils != null) {
                sftpUtils.close();
            }
        }

        // retryable (idempotent) output:
        // return resume(task.dump(), taskCount, control);

        // non-retryable (non-idempotent) output:
        control.run(task.dump());
        return Exec.newConfigDiff();
    }

    @Override
    public ConfigDiff resume(TaskSource taskSource,
            int taskCount,
            FileOutputPlugin.Control control)
    {
        throw new UnsupportedOperationException("sftp output plugin does not support resuming");
    }

    @Override
    public void cleanup(TaskSource taskSource,
            int taskCount,
            List<TaskReport> successTaskReports)
    {
        PluginTask task = taskSource.loadTask(PluginTask.class);

        if (task.getRenameFileAfterUpload()) {
            SftpUtils sftpUtils = null;
            try {
                sftpUtils = new SftpUtils(task);
                for (TaskReport report : successTaskReports) {
                    List<Map<String, String>> moveFileList = report.get(List.class, "file_list");
                    for (Map<String, String> pairFiles : moveFileList) {
                        String temporaryFileName = pairFiles.get("temporary_filename");
                        String realFileName = pairFiles.get("real_filename");

                        sftpUtils.renameFile(temporaryFileName, realFileName);
                    }
                }
            }
            finally {
                if (sftpUtils != null) {
                    sftpUtils.close();
                }
            }
        }
    }

    @Override
    public TransactionalFileOutput open(TaskSource taskSource, final int taskIndex)
    {
        final PluginTask task = taskSource.loadTask(PluginTask.class);
        if (task.getLocalBuffering()) {
            return new SftpLocalFileOutput(task, taskIndex);
        }
        return new SftpRemoteFileOutput(task, taskIndex);
    }
}
